### ISSUE TITLE
_SKYRL_USE_NEW_INFERENCE=1 (default) causes ~15x performance regression vs legacy path

### ISSUE THREAD

**Issue by benrhodes26** [NovaSky-AI/SkyRL#1582]
### Summary

I encountered a severe performance regression when using the new HTTP-based inference path (`_SKYRL_USE_NEW_INFERENCE=1`, the default) compared to the legacy Ray-actor-based path (`_SKYRL_USE_NEW_INFERENCE=0`) for colocated FSDP+vLLM training with tensor parallelism, using the Tinker API entrypoint (`skyrl_train` backend). 

I don't know the root cause, and I'm unsure whether others can easily reproduce this.

**Measured:** Qwen3-30B-A3B on 8x H200 GPUs, colocated TP=8:
- **Legacy path:** ~4-5 min/iteration, generation batches complete in 5-10s
- **New path:** ~67 min/iteration, generation ~20x slower, repeated ~10 min CUDA warmups per turn

The only difference between the two runs was the `_SKYRL_USE_NEW_INFERENCE` environment variable. All other config (model, hardware, parallelism, colocation) was identical.

### Environment

- **SkyRL version:** v0.2.0 (`skyrl_train` backend / Tinker API)
- **vLLM version:** 0.19.0
- **Model:** Qwen/Qwen3-30B-A3B (MoE, 30B total, 3B active)
- **Hardware:** 8x NVIDIA H200 (141 GiB each), NVLink
- **Config:** `colocate_all=true`, `tensor_parallel_size=8`, `enforce_eager=false`, `gpu_memory_utilization=0.5`

### Observations

With the new path, the per-iteration breakdown was approximately:

| Phase | Duration | Notes |
|---|---|---|
| FSDP model load (one-off) | ~5 min | `create_model` took ~304s |
| vLLM server start (one-off) | ~4 min | includes model load + graph compile |
| Per-turn CUDA warmup | ~9-10 min | happens 3x per iteration |
| Per-turn generation | ~12-13 min | 120 requests x ~6s/request |
| Training step | ~67s | fast |

With the legacy path, generation batches of 28-34 samples completed in **8-10 seconds**. CUDA graph compilation happened once at startup (46s) and was not repeated on subsequent wake cycles. Sleep took ~10s, wake(weights) ~1.4s, wake(kv_cache) ~1.0s.

### Additional Observations (maybe related, unclear)

**1. GPU under-utilization on one TP rank (new path only)**

During the slow new-path run, `nvidia-smi dmon` showed one TP worker (GPU 1) consistently at **0% SM utilization** while the other 7 GPUs were at 93-100%. This was not a hardware issue: ECC errors were 0, clocks were normal (1980 MHz), temperature and power were within range. Process-level inspection showed the stalled worker had far fewer voluntary context switches (95K vs 646K for rank 0), suggesting it was CPU-spinning rather than doing GPU work. This was not observed during the legacy path run.

**2. Non-colocated run hits FSDP OOM during model loading**

A non-colocated run with the following config hit a `torch.OutOfMemoryError` during FSDP model initialization:

```
colocate_all=false
policy_num_gpus_per_node=2
num_engines=6
gpu_memory_utilization=0.95
```

This is surprising because the full Qwen3-30B-A3B model is ~60 GiB, well below the 141 GiB H200 capacity, and with `colocate_all=false` the FSDP trainer and vLLM engines should be on separate GPUs. This may be a separate issue, but mentioning here in case there is a common cause (e.g. FSDP not sharding properly, or GPU placement not isolating trainer and inference GPUs correctly).

**Comment by SumanthRH** [comment]
@benrhodes26 Thanks for reporting the issue. It would be good to get a minimal reproduction for the issue. Can you provide the exact commands you used for launching the tinker API server ? Is the tinker client code an off the shelf cookbook example or your own custom code?

**Comment by CharlieFRuan** [comment]
Are you using LoRA by any chance?

**Comment by SumanthRH** [comment]
@CharlieFRuan this is with the tinker entrypoint + FSDP backend so it is with LoRA enabled on inference

**Comment by benrhodes26** [comment]
I am working on a minimal reproduction, but proving difficult. My setup is more complex than a vanilla cookbook recipe, and my initial attempts to distill aren't hitting the 15x slowdown.

I will keep trying though!

**Comment by benrhodes26** [comment]
@SumanthRH @CharlieFRuan : thanks for your patience. I think the attached script should reproduce the regression

[repro_inference_regression.py](https://github.com/user-attachments/files/27277561/repro_inference_regression.py)

(script is obviously vibe coded and messy; hopefully it's enough for you to reproduce. Let me know if you need something cleaner)

**Comment by SumanthRH** [comment]
@benrhodes26 Thanks for sharing the reproduction script!

We've parallelly been able to reproduce the issue! The root cause is in LoRA reloading in vLLM - at every scheduler step, vLLM ends up reloading the LoRA weights when `load_inplace=True` with `AsyncLLM` engine. The same is not seen with the offline `LLM` class. @hao-aaron has a workaround to fix this: #1606. We're also going to make a PR to vLLM to fix the issue soon.
